src/base/octets.rs

```diff
@@ -1073,6 +1190,62 @@ mod serde {
             Ok(value.into())
         }
     }
+
+    #[cfg(feature = "heapless")]
+    pub struct HeaplessVecVisitor<const N: usize>;
+
+    #[cfg(feature = "heapless")]
+    impl<const N: usize> HeaplessVecVisitor<N> {
+        fn new() -> Self {
+            Self
+        }
+
+        pub fn deserialize<'de, D: serde::Deserializer<'de>>(
+            self,
+            deserializer: D,
+        ) -> Result<heapless::Vec<u8, N>, D::Error> {
+            deserializer.deserialize_byte_buf(self)
+        }
+    }
+
+    #[cfg(feature = "heapless")]
+    impl<'de, const N: usize> serde::de::Visitor<'de> for HeaplessVecVisitor<N> {
+        type Value = heapless::Vec<u8, N>;
+
+        fn expecting(&self, f: &mut fmt::Formatter) -> fmt::Result {
+            f.write_fmt(format_args!(
+                "an octet sequence of length {} of shorter",
+                N
+            ))
+        }
+
+        fn visit_borrowed_bytes<E: serde::de::Error>(
+            self,
+            value: &'de [u8],
+        ) -> Result<Self::Value, E> {
+            use core::iter::FromIterator;
+
+            if value.len() > N {
+                return Err(E::invalid_length(value.len(), &self));
+            }
+
+            Ok(heapless::Vec::from_iter(value.iter().copied()))
+        }
+
+        #[cfg(feature = "std")]
```

Comment:
Technically speaking, `heapless` doesn't need `std` to support serialization, but right now `serde/std` is always enabled. This whole visitor implementation might not be necessary, but may come in handy if this changes in the future.

I'm not sure if I could write `#[cfg(feature = "serde/std")]` as I believe that would be technically correct, if possible.

Reply:
I think we can switch to no-std Serde. Let me try that and make a PR.

Reply:
Turns out you can’t enable features on an optional dependency without also enabling the dependency itself. There is something called [weak dependency features](https://doc.rust-lang.org/cargo/reference/unstable.html#weak-dependency-features), but that’s not stable yet.

But, now that I look at it closer: I think you don’t want to implement `visit_byte_buf` or `visit_borrowed_bytes`, anyway. `heapless::Vec` has its own memory into which you have to copy the content, so you gain nothing from taking the reference or the vec. The default impls for both these methods forward to `visit_bytes` – which is what you should implement when you copy the data into your own memory. So, I think you should just implement that.

Reply:
I guess I should have checked the serde docs instead of pattern-matching from a different impl :) Thanks for the good insight.